### PR TITLE
test: AsyncRetryLoop cancel options

### PR DESCRIPTION
**Pull request by dbolduc**
Verify the behavior of this line of code, with a test that would fail if it were removed: https://github.com/googleapis/google-cloud-cpp/blob/10cfef588c819438b85c5a086fd9f44d8f087288/google/cloud/internal/async_retry_loop.h#L194

Also, rearrange the order of the `ConfigureContext` test, so that the `TEST()`s are not broken up by `TEST_F()`s.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10731)

### CODE REVIEW

File: google/cloud/internal/async_retry_loop_test.cc [hunk 1]
```diff
@@ -391,6 +391,47 @@ TEST(AsyncRetryLoopTest, SetsTimeout) {
...
+      "error message");
+
+  // Clear the current options before retrying.
+  OptionsSpan clear(Options{});
+  sequencer.PopFront().set_value(Status(StatusCode::kUnavailable, "try again"));
+  sequencer.PopFront().set_value(0);
+  (void)actual.get();
+}
+
+TEST(AsyncRetryLoopTest, CallOptionsDuringCancel) {
```
Comment (devbww) [inline google/cloud/internal/async_retry_loop_test.cc:417]: Why are `AsyncRetryLoopTest.CallOptionsDuringCancel` and `AsyncRestRetryLoopTest.CallOptionsDuringCancel` subtly different in ways that seem unrelated to REST?  We should strive to make the duplication as obvious as possible.

Indeed perhaps we'd be better off interleaving the `Foo.Bar` and `FooRest.Bar` cases, so that the duplication is localized.

Comment (dbolduc) [reply google/cloud/internal/async_retry_loop_test.cc:417]: > subtly different

Fixed. The differences are just the `*Rest*` invocations / `rest_internal` namespace.

> perhaps we'd be better off interleaving...

Maybe. It is not something I plan to do.

Comment (devbww) [reply google/cloud/internal/async_retry_loop_test.cc:417]: > Fixed. The differences are just the `*Rest*` invocations / `rest_internal` namespace.

It is a shame about the namespaces.  In the bigger picture, we should probably avoid `namespace [foo_]internal { }` wrapping in tests, I reckon.

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/async_retry_loop_test.cc (hunk 1 above)
Comment (dbolduc) [reply google/cloud/internal/async_retry_loop_test.cc:417]: Interesting, but I reckon that using the `*_internal` namespace produces more readable tests in many cases.
